compiler/rustc_serialize/src/opaque.rs

```diff
@@ -38,210 +35,97 @@ pub struct FileEncoder {
-            // room to write the input to the buffer.
-            unsafe {
-                let src = buf.as_ptr();
-                let dst = MaybeUninit::slice_as_mut_ptr(&mut self.buf).add(buffered);
-                ptr::copy_nonoverlapping(src, dst, buf_len);
-            }
-
-            self.buffered = buffered + buf_len;
+        if let Some(dest) = self.buffer_empty().get_mut(..buf.len()) {
+            dest.copy_from_slice(buf);
+            self.buffered += buf.len();
         } else {
-            self.write_all_unbuffered(buf);
+            self.write_all_cold_path(buf);
         }
     }
 
-    fn write_all_unbuffered(&mut self, mut buf: &[u8]) {
-        // If we've already had an error, do nothing. It'll get reported after
-        // `finish` is called.
-        if self.res.is_err() {
-            return;
-        }
-
-        if self.buffered > 0 {
+    /// Write up to `N` bytes to this encoder.
+    ///
+    /// Whenever possible, use this function to do writes whose length has a small and
+    /// compile-time constant upper bound.
+    #[inline]
+    pub fn write_with<const N: usize, V>(&mut self, mut visitor: V)
+    where
+        V: FnMut(&mut [u8; N]) -> usize,
+    {
+        let flush_threshold = const { BUF_SIZE.checked_sub(N).unwrap() };
+        if std::intrinsics::unlikely(self.buffered > flush_threshold) {
             self.flush();
         }
-
-        // This is basically a copy of `Write::write_all` but also updates our
-        // `self.flushed`. It's necessary because `Write::write_all` does not
-        // return the number of bytes written when an error is encountered, and
-        // without that, we cannot accurately update `self.flushed` on error.
-        while !buf.is_empty() {
-            match self.file.write(buf) {
-                Ok(0) => {
-                    self.res = Err(io::Error::new(
-                        io::ErrorKind::WriteZero,
-                        "failed to write whole buffer",
-                    ));
-                    return;
-                }
-                Ok(n) => {
-                    buf = &buf[n..];
-                    self.flushed += n;
-                }
-                Err(ref e) if e.kind() == io::ErrorKind::Interrupted => {}
-                Err(e) => {
-                    self.res = Err(e);
-                    return;
-                }
-            }
-        }
+        // SAFETY: We checked above that that N < self.buffer_empty().len(),
+        // and if isn't, flush ensures that our empty buffer is now BUF_SIZE.
+        // We produce a post-mono error if N > BUF_SIZE.
+        let buf = unsafe { self.buffer_empty().first_chunk_mut::<N>().unwrap_unchecked() };
+        let written = visitor(buf);
+        debug_assert!(written <= N);
+        // We have to ensure that an errant visitor cannot cause self.buffered to exeed BUF_SIZE.
+        self.buffered += written.min(N);
```

Comment:
:shrug: I find both options rather distasteful on account of their code size and `BufReader` does the clamping strategy. What I'd really like is a ranged integer to use here so that the caller can hoist the check. I'm sure LLVM never manages to hoist this out of loops because of the `flush` call.

Normally I'd kick off a perf run for this change, but since there is now a queue I think it's better to not and if this looks bad in the post-merge perf run we can handle it then. Let's be optimistic for now :smile: